feat: more robust oracle interface change check

benesjan opened the pull request:
The oracle interface change check we had was bad as it didn't check the full signature but only the oracle handler names. In this PR I change it to check the full signature.

The code looks too complicated because we need to manually use the typescript parser to extract the method signatures. Unfortunately there doesn't seem to be a simpler way to do this.

mverzilli commented on yarn-project/pxe/src/bin/check_oracle_version.ts:
```diff
@@ -16,6 +20,28 @@ import { ORACLE_INTERFACE_HASH } from '../oracle_version.js';
+ * and the impact of failures here is pretty minimal - it's just an oracle interface change check.
+ *
+ * The code looks too complicated because we need to manually use the typescript parser to extract the method
+ * signatures. There doesn't seem to be a simpler way to do this.
```
we probably don't need to keep this comment for posterity 😅, maybe it belongs more in the pr description than in the code

benesjan replied:
https://github.com/AztecProtocol/aztec-packages/pull/18946/commits/9ac33dd2ed5f11cf834c5136c63924ef7a8c8443

mverzilli commented:
> Unfortunately there doesn't seem to be a simpler way to do this.

Another option could be to define an interface for Oracle in a separate file and just hash that. It would cause some false positives, but I guess that's not a big issue for this use case?

benesjan commented:
* **#18946** <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18946?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.com/github/pr/AztecProtocol/aztec-packages/18946?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `next`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

benesjan commented:
> Another option could be to define an interface for Oracle in a separate file and just hash that. It would cause some false positives, but I guess that's not a big issue for this use case?

Then we would need to keep the file manually in sync with `oracle.ts` which would then result in us forgetting that, changing oracle.ts wide code changes and then not getting this check triggered.

benesjan committed:
feat: more robust oracle interface change check

The oracle interface change check we had was bad as it didn't check the full signature but only the oracle handler names. In this PR I change it to check the full signature.

The code looks too complicated because we need to manually use the typescript parser to extract the method signatures. Unfortunately there doesn't seem to be a simpler way to do this.

mverzilli commented:
> > Another option could be to define an interface for Oracle in a separate file and just hash that. It would cause some false positives, but I guess that's not a big issue for this use case?
> 
> Then we would need to keep the file manually in sync with `oracle.ts` which would then result in us forgetting that, changing oracle.ts wide code changes and then not getting this check triggered.

I meant: 

```typescript
import { type IOracle } from "../somewhere/OracleInterface.ts" // we hash this one

class Oracle implements OracleInterface

/** outside oracle, we don't use Oracle, we use OracleInterface **/
```

benesjan commented:
@mverzilli I see. So you would hash the whole `OracleInterface.ts` file?

That would probably work fine.

Will keep that in mind if this starts misbehaving.
